Common/Securities/Option/StrategyMatcher/OptionPositionCollection.cs

```diff
@@ -0,0 +1,489 @@
+
+            return new OptionPositionCollection(positions, rights, strikes, expirations);
+        }
+
+        /// <summary>
+        /// Slices this collection, returning a new collection containing only
+        /// positions matching the specified <paramref name="comparison"/> and <paramref name="expiration"/>
+        /// </summary>
+        public OptionPositionCollection Slice(BinaryComparison comparison, DateTime expiration, bool includeUnderlying = true)
+        {
+            var expirations = comparison.Filter(_expirations, expiration);
+            if (expirations.IsEmpty)
+            {
+                return includeUnderlying && HasUnderlying ? Empty.Add(UnderlyingPosition) : Empty;
+            }
+
+            var positions = ImmutableDictionary<Symbol, OptionPosition>.Empty;
+            if (includeUnderlying)
+            {
+                OptionPosition underlyingPosition;
+                if (_positions.TryGetValue(Underlying, out underlyingPosition))
+                {
+                    positions = positions.Add(Underlying, underlyingPosition);
+                }
+            }
+
+            var rights = ImmutableDictionary<OptionRight, ImmutableHashSet<Symbol>>.Empty;
+            var strikes = ImmutableSortedDictionary<decimal, ImmutableHashSet<Symbol>>.Empty;
+            foreach (var symbol in expirations.SelectMany(kvp => kvp.Value))
+            {
+                var position = _positions[symbol];
+                positions = positions.Add(symbol, position);
+                rights = rights.Add(symbol.ID.OptionRight, symbol);
+                strikes = strikes.Add(symbol.ID.StrikePrice, symbol);
+            }
+
+            return new OptionPositionCollection(positions, rights, strikes, expirations);
+        }
+
+        public IEnumerable<OptionPosition> ForSymbols(IEnumerable<Symbol> symbols)
+        {
+            foreach (var symbol in symbols)
+            {
+                OptionPosition position;
+                if (_positions.TryGetValue(symbol, out position))
+                {
+                    yield return position;
+                }
+            }
+        }
+
+        public IEnumerable<OptionPosition> ForStrike(decimal strike)
+        {
+            ImmutableHashSet<Symbol> symbols;
+            return _strikes.TryGetValue(strike, out symbols)
+                ? ForSymbols(symbols)
+                : Enumerable.Empty<OptionPosition>();
+        }
+
+        public IEnumerable<OptionPosition> ForExpiration(DateTime expiration)
+        {
+            ImmutableHashSet<Symbol> symbols;
+            return _expirations.TryGetValue(expiration, out symbols)
+                ? ForSymbols(symbols)
+                : Enumerable.Empty<OptionPosition>();
+        }
+
+        /// <summary>
+        /// Deducts the positions contained within the specified <paramref name="match"/>
+        /// </summary>
+        public OptionPositionCollection Accept(OptionStrategyDefinitionMatch match)
```

Comment:
Sweet! This is where all the magic happens.
At this point, we know that all legs in `OptionStrategyDefinitionMatch` are fulfilled?
Wondering if we could differ the adjustment of the positions in the collection -> `positions -= leg.Position;` so it's performed once at the end, similar to the algorithm used by option filtering universe

Reply:
_differ the adjustment of the positions_

After we've found a match we tell the `OptionPositionCollection` (the one in the matcher) to _accept_ the match which simply deducts those positions and returns a new collection instance. The matcher continues to do this until we run out of positions or have gone through all the definitions.

I'm not sure I understand the suggestion here.

Also -- this function was moved to `OptionStrategyDefinitionMatch.RemoveFrom(OptionPositionCollection)` -- it seems presumptuous for `OptionPositionCollection` to have functions accepting matches.